### ISSUE TITLE
NaNs in testing

### ISSUE THREAD

**Issue by UYousafzai** [muditbhargava66/PyxLSTM#26]
### Issue
Getting NaNs on backward propoagation.

### Trace
this was the call1
this was the call12
Batch loss was for batch number {batch_idx}:  tensor(9.6656, device='cuda:0', grad_fn=<NllLossBackward0>)
this was the call13
NaN detected in gradient of embedding.weight
NaN detected in parameter embedding.weight after update
this was the call1
this was the call12
NaN detected in model output
Epoch 1/1, Average Loss: 0.0863
Training completed! Total time: 2.32 seconds

### Code
        print("this was the call1")
        if check_nan(input_seq, "input_seq"):
            break
        
        output, _ = model(input_seq)
        
        print("this was the call12")
        if check_nan(output, "model output"):
            break
        
        output = output.contiguous().view(-1, len(loader_object.vocab))
        target_seq = target_seq.contiguous().view(-1)
        
        loss = criterion(output, target_seq)
        print("Batch loss was for batch number {batch_idx}: ", loss)

        
        print("this was the call13")
        if check_nan(loss, "loss"):
            break

### Dataset Used
gwlms/dewiki-20230701-flair-corpus

            
### Possible things to look at
Note: I am using the language_model.py code provided in the repo with the only change being the dataset I am using.

1. I suspect I might be dealing with exploding gradients. The error "NaN detected in gradient of embedding.weight" is a big clue here. I'm thinking my gradients are probably getting too large during backpropagation.
2. I'm a bit concerned about my loss function. That loss value of 9.6656 seems pretty high to me (the starting with the random dataset you provided was around 7).

### Purpose of Ticket
ruling out the possibility that there might be anything wrong with the way XLSTM has been implemented, additionally would love if you would like to collaborate on getting your implementation to work with a large dataset.

**Comment by UYousafzai** [comment]
### Update
I have possible insights into what might be causing the issue, I have temporarily fixed the issue and got the training to work with the public dataset, however there is a small, a very small possibility this may trace back to the XLSTM implementation (I have to empirically test a few things before I post anymore updates.

**Comment by UYousafzai** [comment]
the issue was resolved, I might do experiments and produce a document on the issue at a future point.

**Comment by muditbhargava66** [comment]
Thank you for the update on the issue. It's great to hear that you've found a temporary fix that allows training to work with the public dataset. I appreciate your proactive approach in investigating the root cause, especially the potential connection to the `xLSTM` implementation.
I look forward to your pull request with the temporary fix. When you're ready, please submit it with:
1. A clear description of the changes made
2. Any relevant test `results` or `logs`
3. A brief explanation of how it resolves the current `issue`
